apps/frontend/src/shared-research/export.ts

```diff
@@ -0,0 +1,102 @@
+import { SearchResponse } from '@algolia/client-search';
+import { isInternalAuthor, ResearchOutputResponse } from '@asap-hub/model';
+import { CsvFormatterStream, Row, format } from '@fast-csv/format';
+import streamSaver from 'streamsaver';
+
+import { GetListOptions } from '../api-util';
+
+type ResearchOutputCSV = Record<
+  keyof Omit<ResearchOutputResponse, 'team'>,
+  string | undefined | boolean
+>;
+
+const caseInsensitive = (a: string, b: string) =>
+  a.localeCompare(b, undefined, { sensitivity: 'base' });
+
+export const researchOutputToCSV = (
+  output: ResearchOutputResponse,
+): ResearchOutputCSV => ({
+  title: output.title,
+  type: output.type,
+  subTypes: output.subTypes.join(','),
+  addedDate: output.addedDate,
+  lastUpdatedPartial: output.lastUpdatedPartial,
+  teams: output.teams
+    .map((team) => team.displayName)
+    .sort(caseInsensitive)
+    .join(','),
+  labs: (output.labs ?? [])
+    .map((lab) => lab.name)
+    .sort(caseInsensitive)
+    .join(','),
+  authors: output.authors
+    .map(
+      (user) =>
+        `${user.displayName}${user.orcid ? ` (${user.orcid})` : ''}${
+          isInternalAuthor(user) ? '' : ' [ext]'
+        }`,
+    )
+    .join(','),
+  tags: output.tags
+    .map((item) => item)
+    .sort(caseInsensitive)
+    .join(','),
+  usedInPublication: output.usedInPublication,
+  sharingStatus: output.sharingStatus,
+  asapFunded: output.asapFunded,
+  link: output.link,
+  doi: output.doi,
+  rrid: output.rrid,
+  accession: output.accession,
+  labCatalogNumber: output.labCatalogNumber,
+  description: output.description,
+  accessInstructions: output.accessInstructions,
+  pmsEmails: output.pmsEmails
+    .map((item) => item)
+    .sort(caseInsensitive)
+    .join(','),
+  publishDate: output.publishDate,
+  id: output.id,
+  created: output.created,
+  lastModifiedDate: output.lastModifiedDate,
+});
+
+export const createCsvFileStream = (
+  csvOptions: Parameters<typeof format>[0],
+  fileName: string,
+) => {
+  const csvStream = format(csvOptions);
+  const fileWriter = streamSaver.createWriteStream(fileName).getWriter();
+  csvStream
+    .on('data', (data) => fileWriter.write(data))
+    .on('end', () => fileWriter.close());
+  return csvStream;
+};
+
+export const MAX_ALGOLIA_RESULTS = 10000;
+
+export const algoliaResultsToStream = async <V>(
+  csvStream: CsvFormatterStream<Row, Row>,
+  getResults: ({
+    currentPage,
+    pageSize,
+  }: Pick<GetListOptions, 'currentPage' | 'pageSize'>) => Readonly<
+    Promise<SearchResponse<V>>
+  >,
+  transform: (result: V) => Row,
+) => {
+  let morePages = true;
+  let currentPage = 0;
+  while (morePages) {
+    // We are doing this in chunks and streams to avoid blob/ram limits.
+    // eslint-disable-next-line no-await-in-loop
+    const data = await getResults({
```

Comment:
how we handle when `getResults` fails?

Reply:
Good catch! we don't and it silently fails. I am asking beatriz how we should handle this

Reply:
![image](https://user-images.githubusercontent.com/1936253/135630753-772aea8f-85d0-4967-b690-f855d757de02.png)